### PR TITLE
JENA-1559: Alternative PrefixMapping implementation; use with TDB1 and TDB2.

### PR DESCRIPTION
**Pull request by afs**

### CODE REVIEW

**Comment by afs** [comment]
The diff may not make this clear but the old `GraphPrefixesProjection` is replaced by `GraphPrefixesProjectionTDB` as it noted "Despite the name "TDB" this is general replacement for DatasetPrefixStorage".

`PrefixMappingBase` is a new `PrefixMapping` base class: `PrefixMappingImpl` has hidden state and its extension mechanism is too limited, causing measurably expensive usage due to [JENA-81|https://issues.apache.org/jira/browse/JENA-81] when working with a lot of prefixes (100+).

The changes in `PrefixMappingImpl` are limited to using `ConcurrentHashMap` instead of `HashMap` - that does not make the class multi-thread-consistent but does address multi-thread access to `HashMap`s.

File: jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingBase.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+import org.apache.jena.shared.PrefixMapping;
+import org.apache.jena.util.SplitIRI;
+
+/**
+ * Framework for implementing {@link PrefixMapping}. It is stateless (unlike
+ * {@code PrefixMappingImpl}) and implements the contract of {@link PrefixMapping},
+ * providing the key algorithm and delegating storage to the subclasses.
+ * <p>
+ * Reverse mapping, looking up a URI to find a prefix is complex. There may be several
+ * possibilities. Applications should not reliy on every implementat being consistent,
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingBase.java]: Typos

- `implementat` this line
- `espeically` on the following line

File: jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingBase.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+ * espeically when writing data when there has to be a choice on on prefix to use to
+ * shorten a URI.
+ * 
+ */
+public abstract class PrefixMappingBase implements PrefixMapping {
+    /* Reverse mappings.
+     * The strict contract of PrefixMapping requires a separate reversemapping to be storoed and manipuated
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingBase.java]: Typos:

- Missing space in `reverse mapping`
- `storoed`
- `manipuated`

File: jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingBase.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+     * for RDF/XML output.
+     * <p>
+     * This is a recurring user question - why does {@code Resource.getNamespace},
+     * {@code Resource.getLocalname} not abbreviate when it is legal Turtle.
+     * <p>
+     * Answer - legacy for RDF/XML.
+     * <p>
+     * See also {@link #qnameFor}.
+     */
+    private void checkLegal(String prefix) {
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingBase.java]: Could this be `public static` so this useful comment is included in Javadoc?

File: jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingOver.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.function.BiConsumer;
+
+import org.apache.jena.iri.IRI;
+import org.apache.jena.riot.system.PrefixMap;
+
+public class PrefixMappingOver extends PrefixMappingBase {
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingOver.java]: Would `PrefixMappingDecorator` be a more descriptive name?

Comment (ajs6f) [reply jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingOver.java]: +1

Comment (afs) [reply jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingOver.java]: Its real name is `PrefixMappingOverPrefixMap` ... seemed a bit long and confusing.

It does not seem to be a decorator.  Would "Adapter" be OK?

Comment (ajs6f) [reply jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingOver.java]: Can't speak for @rvesse, but "Adaptor" is even better to my eye.

File: jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingBase.java (hunk 3 above)
Comment (afs) [reply jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingBase.java]: Done, with named changed to `checkLegalPrefix`.  Typos fixed as well.

**Pull request merged**